drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -58,15 +79,29 @@
     (B)     (E)
     GPIO    GND
 */
-static uint8_t        chan_col_order[LED_MATRIX_COLS] = {0};    // track the channel col order
-static uint8_t        current_row                     = 0;      // LED row scan counter
-static uint8_t        row_idx                         = 0;      // key row scan counter
+#if (DIODE_DIRECTION == COL2ROW)
+static uint8_t chan_col_order[LED_MATRIX_COLS] = {0}; // track the channel col order
+static uint8_t current_row                     = 0;   // LED row scan counter
+static uint8_t current_key_row                 = 0;   // key row scan counter
+#    if (SN32_PWM_CONTROL == SOFTWARE_PWM)
+static uint8_t led_duty_cycle[LED_MATRIX_COLS] = {0}; // track the channel duty cycle
+#    endif
+#elif (DIODE_DIRECTION == ROW2COL)
+/* make sure to `#define MATRIX_UNSELECT_DRIVE_HIGH` in this configuration*/
+static uint8_t      chan_row_order[LED_MATRIX_ROWS_HW] = {0}; // track the channel row order
+static uint8_t      current_key_col                    = 0;   // key col scan counter
+static uint8_t      last_key_col                       = 0;   // key col scan counter
+static matrix_row_t row_shifter                        = MATRIX_ROW_SHIFTER;
+#    if (SN32_PWM_CONTROL == SOFTWARE_PWM)
+static uint8_t      led_duty_cycle[LED_MATRIX_ROWS_HW] = {0}; // track the channel duty cycle
+#    endif
+#endif
 extern matrix_row_t   raw_matrix[MATRIX_ROWS];                  // raw values
 extern matrix_row_t   matrix[MATRIX_ROWS];                      // debounced values
 static matrix_row_t   shared_matrix[MATRIX_ROWS];               // scan values
 static volatile bool  matrix_locked                    = false; // matrix update check
 static volatile bool  matrix_scanned                   = false;
-static const uint32_t periodticks                      = 256;
+static const uint32_t periodticks                      = RGB_MATRIX_MAXIMUM_BRIGHTNESS;
 static const uint32_t freq                             = (RGB_MATRIX_HUE_STEP * RGB_MATRIX_SAT_STEP * RGB_MATRIX_VAL_STEP * RGB_MATRIX_SPD_STEP * RGB_MATRIX_LED_PROCESS_LIMIT);
```

Comment:
```suggestion
static const uint32_t freq                             = (RGB_MATRIX_HUE_STEP * RGB_MATRIX_SAT_STEP * RGB_MATRIX_SPD_STEP * RGB_MATRIX_MAXIMUM_BRIGHTNESS * RGB_MATRIX_LED_PROCESS_LIMIT / 2);
```
seems to be a better performer on `hardware COL2ROW pwm`. Other setups not tested yet, but I have a feeling it's too fast for `software pwm`